### PR TITLE
Add data anonymization module

### PR DESCRIPTION
**Pull request by maks-operlejn-ds**
### Description

The feature for anonymizing data has been implemented. In order to protect private data, such as when querying external APIs (OpenAI), it is worth pseudonymizing sensitive data to maintain full privacy.

Anonynization consists of two steps:

1. **Identification:** Identify all data fields that contain personally identifiable information (PII).
2. **Replacement**: Replace all PIIs with pseudo values or codes that do not reveal any personal information about the individual but can be used for reference. We're not using regular encryption, because the language model won't be able to understand the meaning or context of the encrypted data.

We use *Microsoft Presidio* together with *Faker* framework for anonymization purposes because of the wide range of functionalities they provide. The full implementation is available in `PresidioAnonymizer`.

### Future works

- **deanonymization** - add the ability to reverse anonymization. For example, the workflow could look like this: `anonymize -> LLMChain -> deanonymize`. By doing this, we will retain anonymity in requests to, for example, OpenAI, and then be able restore the original data.
- **instance anonymization** - at this point, each occurrence of PII is treated as a separate entity and separately anonymized. Therefore, two occurrences of the name John Doe in the text will be changed to two different names. It is therefore worth introducing support for full instance detection, so that repeated occurrences are treated as a single object.

### Twitter handle
@deepsense_ai / @MaksOpp

### CODE REVIEW

File: libs/experimental/langchain_experimental/data_anonymizer/utils.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import string
+from faker import Faker
+
+fake = Faker()
+
+# Listed  entities supported by Microsoft Presidio (for now, global and US only)
+# Source: https://microsoft.github.io/presidio/supported_entities/
+pseudoanonymizer_mapping = {
```
Comment (mateusz-wosinski-ds) [inline libs/experimental/langchain_experimental/data_anonymizer/utils.py]: constants are usually written in all capital letters with underscores separating words https://peps.python.org/pep-0008/#constants

File: libs/experimental/langchain_experimental/data_anonymizer/presidio.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+    def __init__(
+        self,
+        analyzed_fields: List[str] = list(pseudoanonymizer_mapping.keys()),
+        language: str = "en",
+        operators: Dict[str, OperatorConfig] = None,
+    ):
+        self.analyzed_fields = analyzed_fields
+        self.language = language
+        self.operators = operators
```
Comment (mateusz-wosinski-ds) [inline libs/experimental/langchain_experimental/data_anonymizer/presidio.py]: please add a comment what is the role of an `operator`

File: docs/extras/use_cases/data_anonymization.ipynb [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+    "\n",
+    "[![Open In Collab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/langchain-ai/langchain/blob/master/docs/extras/use_cases/data_anonymization.ipynb)\n",
+    "\n",
+    "## Use case\n",
+    "\n",
+    "Data anonymization is crucial before passing information to a language model like GPT-4 because it helps protect privacy and maintain confidentiality. If data is not anonymized, sensitive information such as names, addresses, contact numbers, or other identifiers linked to specific individuals could potentially be learned and misused. Hence, by obscuring or removing this personally identifiable information (PII), data can be used freely without compromising individuals' privacy rights or breaching data protection laws and regulations.\n",
+    "\n",
+    "## Overview\n",
+    "\n",
+    "Anoninization consists of two steps:\n",
```
Comment (mateusz-wosinski-ds) [inline docs/extras/use_cases/data_anonymization.ipynb]: Anonimization

File: docs/extras/use_cases/data_anonymization.ipynb [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+    "\n",
+    "1. **Identification:** Identify all data fields that contain personally identifiable information (PII).\n",
+    "2. **Replacement**: Replace all PIIs with pseudo values or codes that do not reveal any personal information about the individual but can be used for reference. We're not using regular encryption, because the language model won't be able to understand the meaning or context of the encrypted data.\n",
+    "\n",
+    "We use *Microsoft Presidio* for anonymization because of the wide range of functionality available - it is implemented as `PresidioAnonymizer`. \n",
```
Comment (mateusz-wosinski-ds) [inline docs/extras/use_cases/data_anonymization.ipynb]: "We use *Microsoft Presidio* together with *Faker* framework for anonymization purposes because of the wide range of functionalities they provide. "

File: docs/extras/use_cases/data_anonymization.ipynb [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+    "sequential_chain(\"You can find our super secret data at https://supersecretdata.com\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Future works\n",
+    "\n",
+    "- **deanonymization** - add the ability to reverse anonymization. For example, the workflow could look like this: `anonymize -> LLMChain -> deanonymize`. In this way, we will retain anonymity in requests to, for example, OpenAI, and then be able restore the original data.\n",
```
Comment (mateusz-wosinski-ds) [inline docs/extras/use_cases/data_anonymization.ipynb]: By doing this, we will retain anonymity in requests to, for example, OpenAI, and then be able restore the original data.

File: docs/extras/use_cases/data_anonymization.ipynb (hunk 4 above)
Comment (mateusz-wosinski-ds) [inline docs/extras/use_cases/data_anonymization.ipynb]: I would personally make it a separate subsection in **Quickstart**

File: libs/experimental/langchain_experimental/data_anonymizer/presidio.py (hunk 2 above)
Comment (eryk-dsai) [inline libs/experimental/langchain_experimental/data_anonymizer/presidio.py]: maybe we could initialize it directly with conditional expression ?
```
self.operators = (
    operators if operators is not None else {
        field: OperatorConfig(
            operator_name="custom", params={"lambda": faker_function}
        )
        for field, faker_function in pseudoanonymizer_mapping.items()
    }
)
```

**Review by eryk-dsai — APPROVED** [review]

File: libs/experimental/pyproject.toml [hunk 6]
```diff
@@ -11,6 +11,9 @@ repository = "https://github.com/langchain-ai/langchain"
 [tool.poetry.dependencies]
 python = ">=3.8.1,<4.0"
 langchain = ">=0.0.239"
+presidio-analyzer = {version = "^2.2.33", optional = true}
```
Comment (piotr-grodek-dsai) [inline libs/experimental/pyproject.toml]: you need to follow-up guidelines for contribution - poetry lock AFAIK should be used as well, otherwise CI fails.

File: libs/experimental/langchain_experimental/data_anonymizer/utils.py
Comment (piotr-grodek-dsai) [inline libs/experimental/langchain_experimental/data_anonymizer/utils.py:1]: Maybe faker_presidio_mapping.py? I feel it is better to have less generic names in such big projects.

File: libs/experimental/langchain_experimental/data_anonymizer/base.py [hunk 7]
```diff
@@ -0,0 +1,13 @@
+from abc import ABC, abstractmethod
+
+
+class AnonymizerBase(ABC):
+    """Base abstract class for anonymizers"""
+
+    def anonymize(self, text: str) -> str:
+        """Anonymize text"""
```
Comment (piotr-grodek-dsai) [inline libs/experimental/langchain_experimental/data_anonymizer/base.py:12]: Likely I would add a short comment about non-virtual interface pattern

**Review by piotr-grodek-dsai — APPROVED** [review]
Left some comments, overall looks like a good start on this module.

File: docs/extras/use_cases/data_anonymization.ipynb (hunk 4 above)
Comment (maks-operlejn-ds) [reply docs/extras/use_cases/data_anonymization.ipynb]: I would say Quickstart is demo and Overview is just a brief description - I followed the way it was implemented in the past in other LangChain notebooks :smile:

File: libs/experimental/tests/unit_tests/test_data_anonymizer.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
+from typing import Dict, List
+
+import pytest
+from presidio_analyzer import PatternRecognizer
+from presidio_anonymizer.entities import OperatorConfig
+
+from langchain_experimental.data_anonymizer import PresidioAnonymizer
+
+
+@pytest.mark.parametrize(
```
Comment (piotr-grodek-dsai) [inline libs/experimental/tests/unit_tests/test_data_anonymizer.py:16]: you need to mark it to be enabled if presidio is installed. You can take a peek on my docker containers tests

Comment (piotr-grodek-dsai) [reply libs/experimental/tests/unit_tests/test_data_anonymizer.py:16]: Probably check should be for faker as well if it is used

File: .github/workflows/langchain_experimental_ci.yml [hunk 9]
```diff
@@ -81,3 +81,35 @@ jobs:
 
       - name: Run tests
         run: make test
+  extended-tests:
```
Comment (piotr-grodek-dsai) [inline .github/workflows/langchain_experimental_ci.yml:84]: It should be independent PR

File: libs/experimental/Makefile [hunk 10]
```diff
@@ -15,6 +15,9 @@ tests:
 test_watch:
 	poetry run ptw --now . -- tests/unit_tests
 
+extended_tests:
```
Comment (piotr-grodek-dsai) [inline libs/experimental/Makefile:18]: It should be in other PR, also likely CONTRIBUTIONS or some docs needs to be updated.
